### ISSUE TITLE
Some #[allow(...)] outer attributes do not apply to just_underscores_and_digits

### ISSUE THREAD

**Issue by pohzipohzi** [rust-lang/rust-clippy#12220]
### Summary

Clippy produces false warnings for the following code. The code is not expected to produce warnings as the  `#[allow(...)]` outer attribute should apply to the following line

```
fn main() {}

fn _just_underscores_and_digits_0() {
    #[allow(clippy::just_underscores_and_digits)]
    let _1 = 1; // warning
}

fn _just_underscores_and_digits_1() {
    #[allow(warnings)]
    let _1 = 1; // warning
}
```

These examples work fine

```
#[allow(clippy::just_underscores_and_digits)]
fn _just_underscores_and_digits_2() {
    let _1 = 1;
}

fn _just_underscores_and_digits_3() {
    #![allow(clippy::just_underscores_and_digits)]
    let _1 = 1;
}
```

`just_underscores_and_digits` can be found in `clippy_lints/src/non_expressive_names.rs`, which houses 3 related lints. One of the other lints (`similar_names`) seem to be a false negative

```
// this should warn but does not (remember to run with `-W clippy::similar_names`)
fn _similar_names() {
    let checked_exp = 1;
    let checked_expr = 1;
}
```

I've skimmed through the code but frankly the more I look the more I wonder if it works at all:
- `levenstein_not_1` doesn't seem to work as expected (try `assert!(!levenstein_not_1("checked_exp", "checked_expr"))`)
- `just_underscores_and_digits` warning stems from `SimilarNamesLocalVisitor`, which has a custom `visit_local` implementation, which may be needed for outer attributes to apply correctly

To be clear, I don't use these lints so they don't bother me, but I did spend good time trying to understand why `#[allow(...)]` wasn't working.  If my hunch is right that these lints aren't working perhaps deprecating them can be an option

### Lint Name

just_underscores_and_digits

### Reproducer

See summary

### Version

```text
rustc 1.77.0-nightly (7ffc697ce 2024-01-24)
binary: rustc
commit-hash: 7ffc697ce10f19447c0ce338428ae4b9bc0c041c
commit-date: 2024-01-24
host: x86_64-unknown-linux-gnu
release: 1.77.0-nightly
LLVM version: 17.0.6
```
